File: validators/rule_schema_validator.py
from __future__ import annotations
import json, re
from dataclasses import dataclass
from typing import Any, Dict, List, Optional, Tuple, Callable

SUPPORTED_TYPES = ("headers","non_empty","range","enum","length","regex","unique","decimal")

@dataclass
class ValidationIssue:
    rule_name: str
    rule_type: str
    path: str
    level: str   # "ERROR" | "WARN"
    message: str

class RuleSchemaValidator:
    """
    Validates & normalizes rules.json. One method per rule type.
    Defaults: decimal.precision=18, decimal.scale=2, exact_scale=False.
    """

    def __init__(
        self,
        id_cols: Optional[List[str]] = None,
        fail_fast: bool = False,
        fail_mode: str = "return",  # "return" | "raise"
    ):
        self.dataset_columns = id_cols
        self.fail_fast = fail_fast
        self.fail_mode = fail_mode
        self.default_precision = 18
        self.default_scale = 2
        # Dispatcher: rule.type -> validator method
        self.validators: Dict[str, Callable[[Dict[str, Any], str, str, str, Optional[List[str]], List[ValidationIssue]], None]] = {
            "headers":   self._validate_headers_rule,
            "non_empty": self._validate_non_empty_rule,
            "range":     self._validate_range_rule,
            "enum":      self._validate_enum_rule,
            "length":    self._validate_length_rule,
            "regex":     self._validate_regex_rule,
            "unique":    self._validate_unique_rule,
            "decimal":   self._validate_decimal_rule,
        }

    # ---------- Public API ----------
    def load(self, json_text: str) -> List[Dict[str, Any]]:
        rules = json.loads(json_text)
        if not isinstance(rules, list):
            raise ValueError("Rules JSON must be a list.")
        return rules

    def load_relaxed(self, json_text: str) -> List[Dict[str, Any]]:
        cleaned = self._strip_json_comments_and_trailing_commas(json_text)
        return self.load(cleaned)

    def validate(self, rules: List[Dict[str, Any]]) -> Tuple[bool, List[Dict[str, Any]], List[ValidationIssue]]:
        """
        Validate rules and return (is_valid, rules, issues).
        
        Returns:
            is_valid: True if no ERROR-level issues found, False otherwise
            rules: List of validated/normalized rules
            issues: List of validation issues (errors and warnings)
            
        If fail_fast=True:
            - fail_mode='return' returns immediately with first error
            - fail_mode='raise' raises ValueError with error details
        """
        issues: List[ValidationIssue] = []
        names_seen = set()
        headers_count = 0

        for idx, rule in enumerate(rules):
            path = f"[{idx}]"
            rtype = str(rule.get("type","")).strip()
            rname = (str(rule.get("name","")).strip() or f"rule_{idx}")
            rule["name"] = rname

            if rtype not in SUPPORTED_TYPES:
                issue = self._err(rname, rtype, path, f"Unsupported type '{rtype}'. Supported: {list(SUPPORTED_TYPES)}")
                issues.append(issue)
                if self.fail_fast:
                    if self.fail_mode == "raise":
                        raise ValueError(f"Rule validation failed: {issue.message}")
                    return False, rules, issues
                continue
            if rname in names_seen:
                issue = self._err(rname, rtype, path, "Duplicate rule name")
                issues.append(issue)
                if self.fail_fast:
                    if self.fail_mode == "raise":
                        raise ValueError(f"Rule validation failed: {issue.message}")
                    return False, rules, issues
            names_seen.add(rname)
            if rtype == "headers": headers_count += 1

            # dispatch to per-rule method
            self.validators[rtype](rule, rname, rtype, path, self.dataset_columns, issues)
            
            # Check for errors after each rule if fail_fast is enabled
            if self.fail_fast:
                errors = [i for i in issues if i.level == "ERROR"]
                if errors:
                    if self.fail_mode == "raise":
                        raise ValueError(f"Rule validation failed: {errors[-1].message}")
                    return False, rules, issues

        # cross-rule advisories
        if headers_count == 0 and self.dataset_columns is not None:
            issues.append(self._warn("<schema>","headers","$", "No 'headers' rule present; column presence not enforced."))
        if headers_count > 1:
            issues.append(self._warn("<schema>","headers","$", "Multiple 'headers' rules present; consider consolidating."))

        # Determine if validation was successful (no errors)
        errors = [i for i in issues if i.level == "ERROR"]
        is_valid = len(errors) == 0
        return is_valid, rules, issues

    # ---------- Per-rule validators (one method per type) ----------
    def _validate_headers_rule(self, rule, rname, rtype, path, dataset_columns, issues):
        self._require_list(rule, "columns", path, rname, rtype, issues)
        # Use instance variable if parameter is None
        cols = dataset_columns if dataset_columns is not None else self.dataset_columns
        if cols:
            missing = [c for c in rule.get("columns", []) if c not in cols]
            if missing:
                issues.append(self._warn(rname, rtype, f"{path}.columns", f"Columns not in dataset hint: {missing}"))

    def _validate_non_empty_rule(self, rule, rname, rtype, path, dataset_columns, issues):
        self._require_list(rule, "columns", path, rname, rtype, issues)
        self._warn_unknown(rule, "columns", dataset_columns, rname, rtype, path, issues)

    def _validate_range_rule(self, rule, rname, rtype, path, dataset_columns, issues):
        self._require_keys(rule, ["column","min","max"], path, rname, rtype, issues)
        self._warn_unknown(rule, "column", dataset_columns, rname, rtype, path, issues)
        if all(k in rule for k in ("min","max")):
            try:
                mn, mx = float(rule["min"]), float(rule["max"])
                if mn > mx: issues.append(self._err(rname, rtype, f"{path}.min/max", "min must be <= max"))
            except Exception:
                issues.append(self._err(rname, rtype, f"{path}.min/max", "min/max must be numeric"))

    def _validate_enum_rule(self, rule, rname, rtype, path, dataset_columns, issues):
        self._require_keys(rule, ["column"], path, rname, rtype, issues)
        self._warn_unknown(rule, "column", dataset_columns, rname, rtype, path, issues)
        # normalize alias
        if "allowed" not in rule and "allowedValues" in rule:
            rule["allowed"] = rule["allowedValues"]
        allowed = rule.get("allowed")
        if not isinstance(allowed, list) or len(allowed) == 0:
            issues.append(self._err(rname, rtype, f"{path}.allowed", "Provide non-empty 'allowed' list"))

    def _validate_length_rule(self, rule, rname, rtype, path, dataset_columns, issues):
        self._require_keys(rule, ["column"], path, rname, rtype, issues)
        self._warn_unknown(rule, "column", dataset_columns, rname, rtype, path, issues)
        mn = int(rule.get("min", 0)); mx = int(rule.get("max", 1_000_000))
        rule["min"], rule["max"] = mn, mx
        if mn < 0 or mx < 0 or mn > mx:
            issues.append(self._err(rname, rtype, f"{path}.min/max", "0 ≤ min ≤ max required"))

    def _validate_regex_rule(self, rule, rname, rtype, path, dataset_columns, issues):
        self._require_keys(rule, ["column","pattern"], path, rname, rtype, issues)
        self._warn_unknown(rule, "column", dataset_columns, rname, rtype, path, issues)
        try:
            re.compile(str(rule.get("pattern","")))
        except re.error as e:
            issues.append(self._err(rname, rtype, f"{path}.pattern", f"Invalid regex: {e}"))

    def _validate_unique_rule(self, rule, rname, rtype, path, dataset_columns, issues):
        self._require_list(rule, "columns", path, rname, rtype, issues)
        self._warn_unknown(rule, "columns", dataset_columns, rname, rtype, path, issues)

    def _validate_decimal_rule(self, rule, rname, rtype, path, dataset_columns, issues):
        self._require_keys(rule, ["column"], path, rname, rtype, issues)
        self._warn_unknown(rule, "column", dataset_columns, rname, rtype, path, issues)
        # defaults
        prec = int(rule.get("precision", self.default_precision))
        scale = int(rule.get("scale", self.default_scale))      # default scale=2
        rule["precision"], rule["scale"] = prec, scale
        rule["exact_scale"] = bool(rule.get("exact_scale", False))
        if prec <= 0 or scale < 0 or scale > prec:
            issues.append(self._err(rname, rtype, f"{path}.precision/scale", "Require precision>0 and 0≤scale≤precision"))
        for b in ("min","max"):
            if b in rule and not self._is_number(rule[b]):
                issues.append(self._err(rname, rtype, f"{path}.{b}", f"'{b}' must be numeric if provided"))

    # ---------- Helpers ----------
    def _require_keys(self, obj: Dict[str,Any], keys: List[str], path: str, rname: str, rtype: str, issues: List[ValidationIssue]):
        for k in keys:
            if k not in obj:
                issues.append(self._err(rname, rtype, f"{path}.{k}", f"Missing required key '{k}'"))

    def _require_list(self, obj: Dict[str,Any], key: str, path: str, rname: str, rtype: str, issues: List[ValidationIssue]):
        val = obj.get(key)
        ok = isinstance(val, list) and len(val) > 0 and all(isinstance(x, str) and x.strip() for x in val)
        if not ok:
            issues.append(self._err(rname, rtype, f"{path}.{key}", f"Provide non-empty list of strings in '{key}'"))

    def _warn_unknown(self, obj, key: str, dataset_columns: Optional[List[str]], rname: str, rtype: str, path: str, issues: List[ValidationIssue]):
        # Use instance variable if dataset_columns parameter is None
        cols = dataset_columns if dataset_columns is not None else self.dataset_columns
        if not cols: return
        if key == "column":
            c = obj.get("column")
            if isinstance(c, str) and c not in cols:
                issues.append(self._warn(rname, rtype, f"{path}.column", f"Column '{c}' not in dataset hint"))
        else:  # list
            lst = obj.get(key, [])
            if isinstance(lst, list):
                unknown = [c for c in lst if c not in cols]
                if unknown:
                    issues.append(self._warn(rname, rtype, f"{path}.{key}", f"Columns not in dataset hint: {unknown}"))

    @staticmethod
    def _is_number(v: Any) -> bool:
        try: float(v); return True
        except Exception: return False

    @staticmethod
    def _err(rule_name, rule_type, path, msg) -> ValidationIssue:
        return ValidationIssue(rule_name, rule_type, path, "ERROR", msg)

    @staticmethod
    def _warn(rule_name, rule_type, path, msg) -> ValidationIssue:
        return ValidationIssue(rule_name, rule_type, path, "WARN", msg)

    @staticmethod
    def _strip_json_comments_and_trailing_commas(s: str) -> str:
        s = re.sub(r"(?m)//.*?$", "", s)           # // line comments
        s = re.sub(r"/\*.*?\*/", "", s, flags=re.S) # /* */ blocks
        s = re.sub(r",\s*([}\]])", r"\1", s)        # trailing commas
        return s

File: java-validators/src/main/java/com/validators/RuleSchemaValidator.java
package com.validators;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.math.BigDecimal;
import java.util.*;
import java.util.regex.Pattern;
import java.util.regex.PatternSyntaxException;

/**
 * Validates and normalizes rule schema JSON for data validation.
 * Supports 8 rule types: headers, non_empty, range, enum, length, regex, unique, decimal.
 * 
 * Defaults: decimal.precision=18, decimal.scale=2, exact_scale=false.
 */
public class RuleSchemaValidator {
    
    private static final Set<String> SUPPORTED_TYPES = Set.of(
        "headers", "non_empty", "range", "enum", "length", "regex", "unique", "decimal"
    );
    
    private final int defaultPrecision;
    private final int defaultScale;
    private final ObjectMapper objectMapper;
    private final Map<String, RuleValidator> validators;
    
    /**
     * Functional interface for rule-specific validation logic.
     */
    @FunctionalInterface
    private interface RuleValidator {
        void validate(Map<String, Object> rule, String rname, String rtype, String path, List<ValidationIssue> issues);
    }
    
    /**
     * Initialize RuleSchemaValidator with default decimal precision/scale and rule type dispatch table.
     */
    public RuleSchemaValidator() {
        this.defaultPrecision = 18;
        this.defaultScale = 2;
        this.objectMapper = new ObjectMapper();
        
        // Dispatcher: rule.type -> validator method
        this.validators = new HashMap<>();
        validators.put("headers", this::validateHeadersRule);
        validators.put("non_empty", this::validateNonEmptyRule);
        validators.put("range", this::validateRangeRule);
        validators.put("enum", this::validateEnumRule);
        validators.put("length", this::validateLengthRule);
        validators.put("regex", this::validateRegexRule);
        validators.put("unique", this::validateUniqueRule);
        validators.put("decimal", this::validateDecimalRule);
    }
    
    // ---------- Public API ----------
    
    /**
     * Load rules from a JSON string. Throws IllegalArgumentException if the root is not a list.
     * 
     * @param jsonText JSON string representing rules
     * @return List of rule maps
     * @throws Exception if JSON parsing fails
     */
    public List<Map<String, Object>> load(String jsonText) throws Exception {
        Object parsed = objectMapper.readValue(jsonText, Object.class);
        if (!(parsed instanceof List)) {
            throw new IllegalArgumentException("Rules JSON must be a list.");
        }
        return objectMapper.readValue(jsonText, new TypeReference<List<Map<String, Object>>>() {});
    }
    
    /**
     * Validate a list of rule dictionaries.
     * Checks for supported types, duplicate names, and delegates to per-type validators.
     * Also checks for cross-rule advisories (e.g., multiple 'headers' rules).
     * 
     * @param rules List of rule maps
     * @return ValidationResult containing validity status and list of issues
     */
    public ValidationResult validate(List<Map<String, Object>> rules) {
        List<ValidationIssue> issues = new ArrayList<>();
        Set<String> namesSeen = new HashSet<>();
        int headersCount = 0;
        
        for (int idx = 0; idx < rules.size(); idx++) {
            Map<String, Object> rule = rules.get(idx);
            String path = "[" + idx + "]";
            
            String rtype = rule.getOrDefault("type", "").toString().trim();
            String rname = rule.getOrDefault("name", "").toString().trim();
            if (rname.isEmpty()) {
                rname = "rule_" + idx;
            }
            rule.put("name", rname);
            
            // Check if type is supported
            if (!SUPPORTED_TYPES.contains(rtype)) {
                issues.add(err(rname, rtype, path, 
                    String.format("Unsupported type '%s'. Supported: %s", rtype, SUPPORTED_TYPES)));
                continue;
            }
            
            // Check for duplicate names
            if (namesSeen.contains(rname)) {
                issues.add(err(rname, rtype, path, "Duplicate rule name"));
            }
            namesSeen.add(rname);
            
            if ("headers".equals(rtype)) {
                headersCount++;
            }
            
            // Dispatch to per-rule method
            RuleValidator validator = validators.get(rtype);
            if (validator != null) {
                validator.validate(rule, rname, rtype, path, issues);
            }
        }
        
        // Cross-rule advisories
        if (headersCount > 1) {
            issues.add(warn("<schema>", "headers", "$", 
                "Multiple 'headers' rules present; consider consolidating."));
        }
        
        // Determine if validation was successful (no errors)
        long errorCount = issues.stream().filter(ValidationIssue::isError).count();
        boolean isValid = errorCount == 0;
        
        return new ValidationResult(isValid, issues);
    }
    
    // ---------- Per-rule validators (one method per type) ----------
    
    /**
     * Validates a 'headers' rule. Requires a non-empty list of column names in 'columns'.
     */
    private void validateHeadersRule(Map<String, Object> rule, String rname, String rtype, String path, List<ValidationIssue> issues) {
        requireList(rule, "columns", path, rname, rtype, issues);
    }
    
    /**
     * Validates a 'non_empty' rule. Requires a non-empty list of column names in 'columns'.
     */
    private void validateNonEmptyRule(Map<String, Object> rule, String rname, String rtype, String path, List<ValidationIssue> issues) {
        requireList(rule, "columns", path, rname, rtype, issues);
    }
    
    /**
     * Validates a 'range' rule. Requires 'column', 'min', and 'max' keys.
     * Checks that min and max are numeric, min <= max, and values are within double bounds.
     * 
     * Valid examples (will pass):
     *   {"type": "range", "column": "age", "min": "0", "max": "120"}
     *   {"type": "range", "column": "price", "min": "0.99", "max": "999.99"}
     *   {"type": "range", "column": "temperature", "min": "-40", "max": "50"}
     *   {"type": "range", "column": "score", "min": "0.0", "max": "100.0"}
     *   {"type": "range", "column": "distance", "min": "1.5e10", "max": "3.0e12"}
     * 
     * Invalid examples (will be rejected):
     *   {"type": "range", "column": "value", "min": "0", "max": "Infinity"}
     *     -> Error: "max value must be a finite number"
     *   {"type": "range", "column": "value", "min": "-Infinity", "max": "100"}
     *     -> Error: "min value must be a finite number"
     *   {"type": "range", "column": "score", "min": "0", "max": "NaN"}
     *     -> Error: "max value must be a finite number"
     *   {"type": "range", "column": "age", "min": "zero", "max": "hundred"}
     *     -> Error: "min/max must be numeric and within valid range"
     *   {"type": "range", "column": "price", "min": "$10", "max": "$100"}
     *     -> Error: "min/max must be numeric and within valid range"
     *   {"type": "range", "column": "age", "min": "65", "max": "18"}
     *     -> Error: "min must be <= max"
     */
    private void validateRangeRule(Map<String, Object> rule, String rname, String rtype, String path, List<ValidationIssue> issues) {
        requireKeys(rule, Arrays.asList("column", "min", "max"), path, rname, rtype, issues);
        
        if (rule.containsKey("min") && rule.containsKey("max")) {
            try {
                double min = Double.parseDouble(rule.get("min").toString());
                double max = Double.parseDouble(rule.get("max").toString());
                
                // Check for reasonable bounds and special values
                if (Double.isInfinite(min) || Double.isNaN(min)) {
                    issues.add(err(rname, rtype, path + ".min", "min value must be a finite number"));
                }
                if (Double.isInfinite(max) || Double.isNaN(max)) {
                    issues.add(err(rname, rtype, path + ".max", "max value must be a finite number"));
                }
                if (min > max) {
                    issues.add(err(rname, rtype, path + ".min/max", "min must be <= max"));
                }
            } catch (NumberFormatException e) {
                issues.add(err(rname, rtype, path + ".min/max", "min/max must be numeric and within valid range"));
            }
        }
    }
    
    /**
     * Validates an 'enum' rule. Requires 'column' key and a non-empty 'allowed' list.
     * Normalizes 'allowedValues' to 'allowed' if present.
     */
    private void validateEnumRule(Map<String, Object> rule, String rname, String rtype, String path, List<ValidationIssue> issues) {
        requireKeys(rule, Collections.singletonList("column"), path, rname, rtype, issues);
        
        // Normalize alias
        if (!rule.containsKey("allowed") && rule.containsKey("allowedValues")) {
            rule.put("allowed", rule.get("allowedValues"));
        }
        
        Object allowed = rule.get("allowed");
        if (!(allowed instanceof List) || ((List<?>) allowed).isEmpty()) {
            issues.add(err(rname, rtype, path + ".allowed", "Provide non-empty 'allowed' list"));
        }
    }
    
    /**
     * Validates a 'length' rule. Requires 'column' key and checks min/max constraints.
     * Sets defaults for min (0) and max (255) if not provided.
     */
    private void validateLengthRule(Map<String, Object> rule, String rname, String rtype, String path, List<ValidationIssue> issues) {
        requireKeys(rule, Collections.singletonList("column"), path, rname, rtype, issues);
        
        int min = 0;
        int max = 255;
        
        try {
            if (rule.containsKey("min")) {
                min = Integer.parseInt(rule.get("min").toString());
            }
            if (rule.containsKey("max")) {
                max = Integer.parseInt(rule.get("max").toString());
            }
        } catch (NumberFormatException e) {
            issues.add(err(rname, rtype, path + ".min/max", "min/max must be valid integers"));
            return;
        }
        
        rule.put("min", min);
        rule.put("max", max);
        
        if (min < 0 || max < 0 || min > max) {
            issues.add(err(rname, rtype, path + ".min/max", "0 ≤ min ≤ max required"));
        }
    }
    
    /**
     * Validates a 'regex' rule. Requires 'column' and 'pattern' keys.
     * Checks that the regex pattern is valid.
     */
    private void validateRegexRule(Map<String, Object> rule, String rname, String rtype, String path, List<ValidationIssue> issues) {
        requireKeys(rule, Arrays.asList("column", "pattern"), path, rname, rtype, issues);
        
        try {
            String pattern = rule.getOrDefault("pattern", "").toString();
            Pattern.compile(pattern);
        } catch (PatternSyntaxException e) {
            issues.add(err(rname, rtype, path + ".pattern", "Invalid regex: " + e.getMessage()));
        }
    }
    
    /**
     * Validates a 'unique' rule. Requires a non-empty list of column names in 'columns'.
     */
    private void validateUniqueRule(Map<String, Object> rule, String rname, String rtype, String path, List<ValidationIssue> issues) {
        requireList(rule, "columns", path, rname, rtype, issues);
    }
    
    /**
     * Validates a 'decimal' rule. Requires 'column' key and checks precision/scale constraints.
     * Sets defaults for precision and scale if not provided. Checks that min/max are numeric if present.
     * 
     * Valid examples (will pass):
     *   {"type": "decimal", "column": "amount", "precision": 10, "scale": 2}
     *   {"type": "decimal", "column": "price", "precision": 18, "scale": 6}
     *   {"type": "decimal", "column": "value", "precision": 10, "scale": 2, "min": -1000, "max": 1000}
     *   {"type": "decimal", "column": "balance", "precision": 15, "scale": 2, "exact_scale": true}
     *   {"type": "decimal", "column": "rate"}  // Uses defaults: precision=18, scale=2
     * 
     * Invalid examples (will be rejected):
     *   {"type": "decimal", "column": "amount", "precision": 0, "scale": 2}
     *     -> Error: "Require precision>0 and 0≤scale≤precision"
     *   {"type": "decimal", "column": "amount", "precision": 10, "scale": -1}
     *     -> Error: "Require precision>0 and 0≤scale≤precision"
     *   {"type": "decimal", "column": "amount", "precision": 5, "scale": 10}
     *     -> Error: "Require precision>0 and 0≤scale≤precision"
     *   {"type": "decimal", "column": "amount", "precision": 20, "scale": 2}
     *     -> Error: "Precision must not exceed 18 (financial standard)"
     *   {"type": "decimal", "column": "amount", "precision": 10, "scale": 8}
     *     -> Error: "Scale must not exceed 6 (common practice)"
     *   {"type": "decimal", "column": "amount", "precision": 10, "scale": 2, "min": "abc"}
     *     -> Error: "'min' must be numeric if provided"
     *   {"type": "decimal", "column": "amount", "precision": 10, "scale": 2, "max": "xyz"}
     *     -> Error: "'max' must be numeric if provided"
     */
    private void validateDecimalRule(Map<String, Object> rule, String rname, String rtype, String path, List<ValidationIssue> issues) {
        requireKeys(rule, Collections.singletonList("column"), path, rname, rtype, issues);
        
        // Defaults
        int precision = defaultPrecision;
        int scale = defaultScale;
        
        try {
            if (rule.containsKey("precision")) {
                precision = Integer.parseInt(rule.get("precision").toString());
            }
            if (rule.containsKey("scale")) {
                scale = Integer.parseInt(rule.get("scale").toString());
            }
        } catch (NumberFormatException e) {
            issues.add(err(rname, rtype, path + ".precision/scale", "precision/scale must be valid integers"));
            return;
        }
        
        rule.put("precision", precision);
        rule.put("scale", scale);
        
        // Parse exact_scale with validation
        boolean exactScale = false;
        if (rule.containsKey("exact_scale")) {
            Object exactScaleValue = rule.get("exact_scale");
            if (exactScaleValue instanceof Boolean) {
                exactScale = (Boolean) exactScaleValue;
            } else {
                String strValue = exactScaleValue.toString().toLowerCase();
                if ("true".equals(strValue)) {
                    exactScale = true;
                } else if (!"false".equals(strValue)) {
                    issues.add(err(rname, rtype, path + ".exact_scale", "exact_scale must be boolean (true/false)"));
                    return;
                }
            }
        }
        rule.put("exact_scale", exactScale);
        
        if (precision <= 0 || scale < 0 || scale > precision) {
            issues.add(err(rname, rtype, path + ".precision/scale", 
                "Require precision>0 and 0≤scale≤precision"));
        }
        if (precision > 18) {
            issues.add(err(rname, rtype, path + ".precision", 
                "Precision must not exceed 18 (financial standard)"));
        }
        if (scale > 6) {
            issues.add(err(rname, rtype, path + ".scale", 
                "Scale must not exceed 6 (common practice)"));
        }
        
        for (String bound : Arrays.asList("min", "max")) {
            if (rule.containsKey(bound) && !isNumber(rule.get(bound))) {
                issues.add(err(rname, rtype, path + "." + bound, 
                    String.format("'%s' must be numeric if provided", bound)));
            }
        }
    }
    
    // ---------- Helpers ----------
    
    /**
     * Helper to ensure required keys are present in a rule object.
     * Adds an error to issues for each missing key.
     */
    private void requireKeys(Map<String, Object> obj, List<String> keys, String path, 
                            String rname, String rtype, List<ValidationIssue> issues) {
        for (String key : keys) {
            if (!obj.containsKey(key)) {
                issues.add(err(rname, rtype, path + "." + key, 
                    String.format("Missing required key '%s'", key)));
            }
        }
    }
    
    /**
     * Helper to ensure a key in a rule object is a non-empty list of strings.
     * Adds an error to issues if not satisfied.
     */
    private void requireList(Map<String, Object> obj, String key, String path, 
                            String rname, String rtype, List<ValidationIssue> issues) {
        Object val = obj.get(key);
        boolean ok = val instanceof List 
            && !((List<?>) val).isEmpty() 
            && ((List<?>) val).stream()
                .allMatch(x -> x instanceof String && !((String) x).trim().isEmpty());
        
        if (!ok) {
            issues.add(err(rname, rtype, path + "." + key, 
                String.format("Provide non-empty list of strings in '%s'", key)));
        }
    }
    
    /**
     * Returns true if v can be converted to a number, false otherwise.
     */
    private static boolean isNumber(Object v) {
        try {
            new BigDecimal(v.toString());
            return true;
        } catch (Exception e) {
            return false;
        }
    }
    
    /**
     * Helper to create an ERROR-level ValidationIssue.
     */
    private static ValidationIssue err(String ruleName, String ruleType, String path, String msg) {
        return new ValidationIssue(ruleName, ruleType, path, "ERROR", msg);
    }
    
    /**
     * Helper to create a WARN-level ValidationIssue.
     */
    private static ValidationIssue warn(String ruleName, String ruleType, String path, String msg) {
        return new ValidationIssue(ruleName, ruleType, path, "WARN", msg);
    }
    
    /**
     * Result object containing validation status and issues.
     */
    public static class ValidationResult {
        private final boolean valid;
        private final List<ValidationIssue> issues;
        
        public ValidationResult(boolean valid, List<ValidationIssue> issues) {
            this.valid = valid;
            this.issues = Collections.unmodifiableList(issues);
        }
        
        public boolean isValid() {
            return valid;
        }
        
        public List<ValidationIssue> getIssues() {
            return issues;
        }
        
        public List<ValidationIssue> getErrors() {
            return issues.stream()
                .filter(ValidationIssue::isError)
                .toList();
        }
        
        public List<ValidationIssue> getWarnings() {
            return issues.stream()
                .filter(ValidationIssue::isWarning)
                .toList();
        }
    }
}
